File: launcher/ui/ConfigListWidget.py
import sys

import PySide2.QtCore
import PySide2.QtWidgets
from PySide2.QtGui import QGuiApplication
from PySide2.QtCore import QStringListModel, Slot,QModelIndex
from PySide2.QtWidgets import QWidget, QVBoxLayout, QGroupBox, QHBoxLayout, QPushButton, QListView, QApplication,QFileDialog
import configparser

from launcher.config import RECOLL_CONFIG_FILE


class ConfigListWidget(QGroupBox):

    _confParser=configparser.ConfigParser()
    _confParser.optionxform=str
    with open(RECOLL_CONFIG_FILE) as f:
        file_content = '[DEFAULT]\n' + f.read()
    _confParser.read_string(file_content)
    def __init__(self, parent=None, showGroupName: str = None, section="DEFAULT",
                 groupName: str = None):
        super().__init__(showGroupName, parent)
        self._groupName = groupName
        self._section = section
        self.init_ui()

        self._confParser = None
        # TODO exception
        self.init_data()
        self._init_conn()

    def _init_conn(self):
        self._btnAdd.clicked.connect(self._btn_add)
        self._btnDelete.clicked.connect(self._btn_delete)
        pass

    def _btn_add(self):
        path=QFileDialog.getExistingDirectory(self,caption="Select Dir",dir="~")
        slist=self._listmodel.stringList()
        slist.append(path)
        self._listmodel.setStringList(slist)
        pass
    def _btn_delete(self):
        select=self._listView.selectedIndexes()
        for item in select:
            self._listmodel.removeRow(item.row())


    # @classmethod
    def save_values(self):
        row=""
        #TODO remove duplicated
        for i in range(self._listmodel.rowCount()):
            row += "\t"+self._listmodel.index(i).data()+"\\\n"

        row=row[:-2]
        print(row)
        # self._confParser[self._section][self._groupName]=row
        ConfigListWidget._confParser.set(self._section,self._groupName,row)
        with open(RECOLL_CONFIG_FILE,'w+') as conf:
            ConfigListWidget._confParser.write(conf)
            conf.seek(0)
            lines=conf.readlines()

        with open(RECOLL_CONFIG_FILE,'w') as conf:
            for i in lines[1:]:
                conf.write(i)


    def init_data(self):
        # TODO default values
        self._listmodel = QStringListModel(
            str(ConfigListWidget._confParser[self._section][self._groupName]).replace("\\", "").strip().replace("\t","").splitlines())
        self._listView.setModel(self._listmodel)


    def init_ui(self):
        self._vlayout = QVBoxLayout()
        self.setLayout(self._vlayout)
        self._listView = QListView()
        self._hlayout = QHBoxLayout()

        self._vlayout.addWidget(self._listView)
        self._vlayout.addLayout(self._hlayout)

        self._btnAdd = QPushButton("Add")
        self._btnDelete = QPushButton("Delete")

        self._hlayout.addWidget(self._btnAdd)
        self._hlayout.addWidget(self._btnDelete)


if __name__ == "__main__":
    app = QApplication()
    widget = ConfigListWidget(showGroupName="Topdir", confName="/home/tender/.config/everylauncher/recoll/recoll.conf",
                              groupName="topdirs")
    widget.show()

    sys.exit(app.exec_())

File: src/configlistwidget.cpp
#include "configlistwidget.h"
#include "config.h"

#include <QDir>
#include <QFile>
#include <QStandardPaths>
#include <QTextStream>
#include <QVBoxLayout>
#include <QDebug>
#include <QProcess>
#include <QFileDialog>

ConfigListWidget::ConfigListWidget(QString showGroupName, QString section,
                                   QWidget *parent)
    :QGroupBox (showGroupName,parent) ,mgroupName(showGroupName), msection(section) {
  // TODO conf Parser
  btnAdd = new QPushButton(tr("Add"));
  btnDelete = new QPushButton(tr("Delete"));
  listView = new QListView(this);
  listmodel=new QStringListModel(this);

  init_ui();
  init_data();
  init_conn();
}

void ConfigListWidget::initConfigFile()
{
  auto filedir =QDir(
      QStandardPaths::writableLocation(QStandardPaths::AppConfigLocation));
  filedir.cd(RECOLL_CONFIG_DIR);
  auto filepath = filedir.absoluteFilePath(RECOLL_CONFIG_FILE);
  if(confFile!=nullptr){
      confFile->close();
      delete confFile;
      confFile=nullptr;
  }
  confFile = new QFile(filepath);

  msettings.clear();
  if (confFile->exists()) {
      confFile->open(QFile::ReadWrite);
      if(qts==nullptr)
        qts=new QTextStream(confFile);
      while (!qts->atEnd()) {
          auto line=qts->readLine();
          while(line.endsWith('\\')){
//              line.replace('\\',' ');
              line+=qts->readLine();
          }
          auto tmp=line.trimmed().split("=");
          if(tmp.length()!=2){
              continue;
          }
          if(tmp[1].trimmed().isEmpty()){
              continue;
          }
          msettings[tmp[0]]=tmp[1];
//          auto k=tmp[0];
//          auto v=tmp[1];
      }
//   confFile->reset();

      if(!confFile->resize(0)){
          qDebug()<<"resize failed";
      }
      confFile->flush();
      qts->seek(0);

  }
}

void ConfigListWidget::init_ui() {
  auto vlayout = new QVBoxLayout();
  this->setLayout(vlayout);
  auto hlayout = new QHBoxLayout();

  hlayout->addWidget(this->btnAdd);
  hlayout->addWidget(this->btnDelete);

  vlayout->addWidget(this->listView);
  vlayout->addLayout(hlayout);
}

void ConfigListWidget::init_data()
{
//    qDebug()<<this->confgSettings->allKeys();
    auto items=msettings[msection].split("\\").replaceInStrings("\t","");
    this->listmodel->setStringList(items);
    this->listView->setModel(this->listmodel);

}

void ConfigListWidget::init_conn() {
    connect(this->btnAdd, &QPushButton::clicked, this,&ConfigListWidget::btn_add);
    connect(this->btnDelete, &QPushButton::clicked, this,&ConfigListWidget::btn_delete);
}

void ConfigListWidget::btn_add()
{

    auto path=QFileDialog::getExistingDirectory(this,tr("Add Monitor Floder"),QDir::homePath());
    if(this->listmodel->insertRow(this->listmodel->rowCount())){
        auto index=this->listmodel->index(this->listmodel->rowCount()-1,0);
        this->listmodel->setData(index,path);

    }
}
void ConfigListWidget::btn_delete()
{
    auto selectmodel=this->listView->selectionModel();
    for(auto item:selectmodel->selectedRows()){
        listmodel->removeRow(item.row());

    }



}

void ConfigListWidget::save_config()
{
   auto lm=this->listmodel->stringList();
//   qts->setDevice(confFile);
   *qts<<"\n";
   QString line;
   for(auto item:lm){
       if(item.trimmed().isEmpty())
           continue;
       line+='\t'+item+"\\\n";
   }
   line.replace(line.length()-2,2,' ');
//   ts.seek(0);
   *qts<<this->msection<<"="<<line;
   qts->flush();

}
QTextStream *ConfigListWidget::qts=nullptr;

QFile *ConfigListWidget::confFile=nullptr;
QMap<QString,QString> ConfigListWidget::msettings;
